### PR TITLE
Linter with no_allow rule

### PR DESCRIPTION
**Pull request by flyinghermit**
This PR implements a linter functionality with one `no_allow` lint rule. For high-level overview on linter, refer to [Lint RFD PR](https://github.com/gravitational/predicate-lang/pull/62).

## Details:
- Adds a `lint` package and a CLI command that can be invoked as `$ predicate lint <policy file or directory`.
- `Lint rule`: Lint functionality implemented within predicate as a python class. This PR implements `no_allow` lint rule to demonstrate linting functionality. 
- `no_allow`: Lint rule which enables policy administrator to forbid certain rules to be added to `allow` rule. When a new policy is created with matching forbidden rule, running linter will verify the rule using `Predicate.equivalent` and report back with details. Only `allow` rule is evaluated. 
- `Lint rule file`: Rule file that extends `Lint rule` with configurable rules. Some `Lint rule` such as detecting two duplicate policies can be self-contained (no additional configuration needed) while others, such as `no_allow` rule, requires an additional rule file. In practice, users will only need to update rule files and lint config file. Check out sample [`no_allow` rule file](https://github.com/gravitational/predicate-lang/blob/39fe11b33de22a8bd09b0620ecaaffc1d4260312/predicate/lint/tests/data/no_allow.py).
- `Lint config file`: Config file `predicatelint.yml` with active lint rules, rule files path, etc. Any configuration to update linter runtime will go here. Currently lint command expects this config file in the root directory of predicate. Here's a working [config file sample](https://github.com/gravitational/predicate-lang/blob/39fe11b33de22a8bd09b0620ecaaffc1d4260312/predicate/predicatelint.yml).
- Lint results are printed to stdout. JSON output is also supported.

## Testing:

- Pull this branch and run: 
```
# test single policy
$ predicate lint examples/access.py

# test policy directory
$ predicate lint examples

# JSON output
$ predicate lint examples/access.py --out=json

``` 

- Testing new `no_allow` rule:
    - First update the  [`no_allow` rule file](https://github.com/gravitational/predicate-lang/blob/39fe11b33de22a8bd09b0620ecaaffc1d4260312/predicate/lint/tests/data/no_allow.py) with new rule. 
    - Create policy file with violating allow rule. 
    - Run linter on this new policy file as `$ predicate lint <new policy file>`

## Adding entirely new lint rule:
The process to add entirely new lint rule would be:
- Write a new rule class in `lint/rule.py`.
- Update `predicatelint.yaml`.
- Update `Linter` class in `lint/linter.py` to add new lint rule class.

### CODE REVIEW

File: predicate/README.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ class Teleport:
                 | (User.traits["team"] == ("admins",))
             ),
         ),
-        options=OptionsSet(Options((Options.max_session_ttl < Duration.new(hours=10)))),
```
Comment (xacrimon) [inline predicate/README.md:32]: Thanks! I missed these

File: predicate/cli/tests/data/buggy_policy.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+class Developer:
+    p = Policy(
+        name="developer",
+        loud=False,
+        allow=Rules(
+            AccessNode(
+                ((AccessNode.login == "testuser") & (User.name != "testuser"))
+                | (User.traits["team"] == ("testteam",))
+            ),
+            Rules(
```
Comment (xacrimon) [inline predicate/cli/tests/data/buggy_policy.py:14]: Is this faulty rules node supposed to be here for the test functionality? referring to the fact it's nested within another rules node, don't think we support this

File: predicate/lint/linter.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                # collect policy files except the ones that trails with __
+                # filters files or directories such as __init__, __pycache__ etc.
+                if "__" not in file.stem:
+                    self.policies_filepath.append(str(file))
+        else:
+            self.policies_filepath.append(policy_file_path)
+
+    def get_lint_config(self):
+        """Returns linter config file"""
+        # TODO: remove hardcoded config file location once desired config location path is decided.
```
Comment (xacrimon) [inline predicate/lint/linter.py:56]: @klizhentas thoughts on this? iirc most linters for modern ecosystems like Rust/JS/TS do this by starting at the current working dir, searching for this file. If not found, go to the parent, rinse and repeat until you hit a permission error or the root.

File: predicate/cli/tests/data/buggy_policy.py (hunk 2 above)
Comment (flyinghermit) [reply predicate/cli/tests/data/buggy_policy.py:14]: Yup this one is to test faulty rule.

File: predicate/lint/linter.py (hunk 3 above)
Comment (flyinghermit) [reply predicate/lint/linter.py:56]: I created a ticket to track this - https://github.com/gravitational/predicate-lang/issues/79 and will resolve in another PR.

**Review by xacrimon — APPROVED** [review]

**Pull request merged**